### DISCUSSION TITLE
Multiple accounts in Outlook, how to create multiple signature files ?

### DISCUSSION THREAD

**Discussion by Imaginos** [Set-OutlookSignatures/Set-OutlookSignatures#64] [Q&A] [answered]
Hello Markus
Thank you again for your great work !

I have a question : 
I have 2 standard mailboxes in my Outlook (2013).
When I run the script, It finds the 2 accounts, but, there is only one signature file set created named (signature.*)
Is it possible to create 2 set of files, ie 'login1_signature.*' and 'login2_signature.*' ?

I use the param SignaturesForAutomappedAndAdditionalMailboxes, of course.
I've read the docs but can't find the answer.

**Comment by GruberMarkus** [comment]
Every centrally stored signature template is only applied to the mailbox with the highest priority allowed to use it. This ensures that no mailbox with lower priority can overwrite a signature intended for a higher priority mailbox. See 'Signature and OOF application order' in 'README' for details.

Workarounds:
- You can create multiple template files with different names. You can then assign each template to a different group or e-mail address in your ini file. See 'Allowed tags' in 'README' for details. The sample file '_signatures.ini' also has examples for these two cases.
- You can use the same template file multiple times in your ini and set a different OutlookSignatureName for each entry. See 'How to work with ini files' in 'README' for details and examples (search for 'Solution B: Use only one template (possible with v3.1.0 and newer)').

Remember: You can deploy signatures for auto-mapped and additional mailboxes, but you cannot set a default signature for these types of mailboxes.

I hope that helps.

**Comment by Imaginos** [comment]
No, it doesn't really help, and I still don't understand why the script lists my 2 mailboxes (with no shared box) but cannot generate one signature per account.
What's the point of listing all the accounts, then ?

Using ini cannot be used, I don't know how many times it happens and for which accounts in the company.

**Comment by GruberMarkus** [comment]
I might not fully understand what you are trying to achieve and which exact problem you encounter.

Please provide the full output of Set-OutlookSignatures and describe your goal and the problem by referring to the script output.

**Comment by Imaginos** [comment]
OK, I know sometimes it's hard for me to be clear, I must be more explicit.

My main user is François Pignon (test user, logged in windows, main user of outlook).
In Outlook, I then added my account : Tom MyName
Outlook is restarted to ensure all is recorded.

The script is started.
I then have the "reponse.html" and "signature.html" filled with Pignon's data.
I would have the script create something like : 
- f.pignon_signature.html
- f.pignon_reponse.html
- t.MyName_signature.html
- f.MyName_reponse.html

Log file attached.
Thanx for looking at this :)

[Signatures.log](https://github.com/GruberMarkus/Set-OutlookSignatures/files/10268612/Signatures.log)

**Comment by GruberMarkus** [comment]
Don't worry, everything is fine. I guess English is a second language to you as well as it is to me.

I think I understand your situation now: You want to have a personalized signature for each mailbox, where each signature should have a separate name based on some attribute of each mailbox.

Remark: This probably only makes sense if you use at least on replacement variable from the `$CURRENTMAILBOX[...]$` range, and not only replacement variables from the `$CURRENTUSER[...]$` range.
<br><br>  
At the current stage of development, this is possible if you have a list of all mailboxes. You can then create a separate entry for each mailbox in your INI file, using the same template name but a different Outlook signature name.
An example:
```
[signature.htm]
f.pignon@example.com
defaultNew
OutlookSignatureName = f.pignon_signature

[signature.htm]
t.myname@example.com
defaultNew
OutlookSignatureName = t.myname_signature

# one more [signature.htm] entry for each mailbox

[reponse.htm]
f.pignon@example.com
defaultReplyForward
OutlookSignatureName = f.pignon_reponse

[reponse.htm]
t.myname@example.com
defaultReplyForward
OutlookSignatureName = t.myname_reponse
```
I know that this option is used by several customers to only have to maintain one template for many shared mailboxes.  
The maximum I have seen until now are about 500 mailboxes sharing one template but having an individual signature name for each mailbox.
<br><br>
If you do not want to maintain these INI entries, Set-OutlookSignatures would need to be extended, for example by programming a new feature as "Dynamic signature names by supporting replacement variable for OutlookSignatureName":
```
[reponse.htm]
defaultReplyForward
OutlookSignatureName = $CURRENTMAILBOXMAIL$_reponse
```
At the moment, I don't have this feature on my roadmap, and I can't say off the top of my head whether I'd consider it useful for a broad user base.

**Comment by Imaginos** [comment]
Hi

I try but it doesn't work
I have the files 
- Signature.htm
- t.MyName_Signature.htm

But the files are the same.
I mean both files are filled with Pignon's info.

**Comment by GruberMarkus** [comment]
Make sure you do not use CURRENTUSER replacement variables, but CURRENTMAILBOX replacement variables.

**Comment by Imaginos** [comment]
I didn't use the variable, i tested the ini file modification : 

[Signature.docx]
defaultNew
OutlookSignatureName = Signature

[Signature.docx]
t.MyName@company.fr
defaultNew
OutlookSignatureName = t.myname_Signature

**Comment by GruberMarkus** [comment]
The feature works fine in all my test environments, but there could still be a bug or misconfiguration.

Please start Set-OutlookSignatures with the '-verbose' parameter and provide the output.

If possible, do not change the log file content. If you obfuscate personal data in the log file, please keep the original log file, as I may come back to you with some questions that will require you to convert the obfuscated data back to the original data in order to answer.

**Comment by Imaginos** [comment]
Environnement : 
docx, ini and replacement ps1 are located on a windows server share.
Script is executed on a win10 pro (french)

The INI file : 
[<Set-OutlookSignatures configuration>]
SortOrder = Ascending

[Signature.docx]
defaultNew
OutlookSignatureName = Signature

[Signature.docx]
t.hero@lanono.fr
defaultNew
OutlookSignatureName = Signature_t.hero

[Reponse.docx]
defaultReplyFwd

And the log file with the verbose param : 
Changes made are systemic on my name and the name of the company.
I still have the original if needed, but changes shouldn't be a problem.
[Signatures - Copie.log](https://github.com/GruberMarkus/Set-OutlookSignatures/files/10285005/Signatures.-.Copie.log)

**Comment by GruberMarkus** [comment]
From a log file perspective, everything works as configured in the INI file:
```
Mailbox t.hero@lanono.fr @2022-12-22T09:14:57+01:00@
  Get group membership of mailbox @2022-12-22T09:14:57+01:00@
    [...]
  Get SMTP addresses @2022-12-22T09:14:58+01:00@
    [...]
  Get data for replacement variables @2022-12-22T09:14:58+01:00@
    [...]
  Process common templates @2022-12-22T09:14:58+01:00@
    'Reponse.docx' (ini index #3) @2022-12-22T09:14:58+01:00@
      [...]
      Template already processed before with higher priority, no need to update signature
      [...]
    'Signature.docx' (ini index #1) @2022-12-22T09:14:58+01:00@
      [...]
      Template already processed before with higher priority, no need to update signature
      [...]
  Process group specific templates @2022-12-22T09:14:58+01:00@
  Process mailbox specific templates @2022-12-22T09:14:58+01:00@
    'Signature.docx' (ini index #2) @2022-12-22T09:14:58+01:00@
      Check permissions
        Allows
          First e-mail address match: t.hero@lanono.fr
        Denies
          Group: Mailbox is not member of any denied group
          E-Mail address: Mailbox does not have any denied e-mail address
        Use template as there is at least one allow and no deny for this mailbox
      Outlook signature name: 'Signature_t.hero'
      Create temporary file copy
      Replace picture variables
      Replace non-picture variables
      Export to HTM format
      Export to RTF format
      Export to TXT format
      Embed local files in HTM format and add marker
      Copy signature files to 'C:\Users\f.pignon\AppData\Roaming\Microsoft\Signatures'
      Remove temporary files
      Set signature as default for new messages (Outlook profile 'Outlook')
```
<br>
Does the file `C:\Users\f.pignon\AppData\Roaming\Microsoft\Signatures\Signature_t.hero.htm` exist?

**Comment by GruberMarkus** [comment]
Addendum: If you want mailbox specific variables in `C:\Users\f.pignon\AppData\Roaming\Microsoft\Signatures\Signature_t.hero.htm` and `C:\Users\f.pignon\AppData\Roaming\Microsoft\Signatures\Signature.htm`, you need to use the `$CURRENTMAILBOX[...]` replacement variables in your templates for the mailbox specific attributes.

Be aware that you use a custom replacement variable configuration file which does not define any mailbox specific replacement variables, only replacement variables based on the currently logged in user.

The default replacement variables delivered with the script are not available because your custom replacement variable config file does not import the default file delivered with Set-OutlookSignatures. The following content from the default file `default replacement variables.ps1` describes how you can load default replacement variables and overwrite or extend them with custom replacement variables:
```
# The following steps are recommended:
# 1. Create a new custom configuration file in a separate folder.
# 2. The first step in the new custom configuration file should be to load the default configuration file:
#    # Loading default replacement variables shipped with Set-OutlookSignatures
#    . ([System.Management.Automation.ScriptBlock]::Create((Get-Content -LiteralPath $(Join-Path -Path $(Get-Location).path -ChildPath '\config\default replacement variables.ps1') -Raw)))
# 3. After importing the default configuration file, existing replacement variables can be altered with custom definitions and new replacement variables can be added.
# 4. Instead of altering existing replacement variables, it is recommended to create new replacement variables with modified content.
# 5. Start Set-OutlookSignatures with the parameter 'ReplacementVariableConfigFile' pointing to the new custom configuration file.
```

**Comment by Imaginos** [comment]
I don't understand how to load the default file first.
When shall I use the '([System.Management.Automation.ScriptBlock]:...' command ?

**Comment by GruberMarkus** [comment]
`\\nono2000.local\NETLOGON\Set-OutlookSignatures\config\replacement variables.ps1`:

```
# Loading default replacement variables shipped with Set-OutlookSignatures
. ([System.Management.Automation.ScriptBlock]::Create((Get-Content -LiteralPath $(Join-Path -Path $(Get-Location).path -ChildPath '\config\default replacement variables.ps1') -Raw)))


# Currently logged in user
$ReplaceHash['$PRENOM$'] = [string]$ADPropsCurrentUser.givenname
$ReplaceHash['$NOM$'] = [string]$ADPropsCurrentUser.sn
$ReplaceHash['$SERVICE$'] = [string]$ADPropsCurrentUser.department
$ReplaceHash['$FONCTION$'] = [string]$ADPropsCurrentUser.title
$ReplaceHash['$CURRENTUSERTELEPHONE$'] = [string]$ADPropsCurrentUser.telephonenumber
$ReplaceHash['$CURRENTUSERMOBILE$'] = [string]$ADPropsCurrentUser.mobile
$ReplaceHash['$EMAIL$'] = [string]$ADPropsCurrentUser.mail
$ReplaceHash['$NOMCOMPLET$'] = [string]$ADPropsCurrentUser.displayname
EN CLAIR :
# Manager of currently logged in user
$ReplaceHash['$CURRENTUSERMANAGERMAIL$'] = [string]$ADPropsCurrentUserManager.mail
$ReplaceHash['$CURRENTUSERMANAGERDISPLAYNAME$'] = [string]$ADPropsCurrentUserManager.displayname
EN CLAIR :
$ReplaceHash['$TELEPHONE$'] = $(if (-not $ReplaceHash['$CURRENTUSERTELEPHONE$']) { '' } else { "Tél : " + $ReplaceHash['$CURRENTUSERTELEPHONE$']} )
$ReplaceHash['$LIGNEDIRECTE$'] = $(if (-not $ReplaceHash['$CURRENTUSERMOBILE$']) { '' } else { "Ligne directe : " + $ReplaceHash['$CURRENTUSERMOBILE$']} )
if ($ReplaceHash['$CURRENTUSERMANAGERGIVENNAME$'] -and $ReplaceHash['$CURRENTUSERMANAGERMAIL$']){
	$ReplaceHash['$MANAGER$'] = "En cas d'urgence, vous pouvez contacter " + $ReplaceHash['$CURRENTUSERMANAGERDISPLAYNAME$'] + " à l'adresse : " + $ReplaceHash['$CURRENTUSERMANAGERMAIL$'] + "."
}else{
	$ReplaceHash['$MANAGER$'] =''
}
```

You than have all default variables available which are described in the `Variable replacement` chapter in the `Readme` file.

**Comment by Imaginos** [comment]
I added the lines to my replacement file but it doesn't change anything.
But I found the trick !
In my replacement, i changed $ADPropsCurrentUser by $ADPropsCurrentMailbox

And then, all is fine !
The files are correctly created and it seems they are correctly assigned to the accounts.

Thank you !

**Comment by GruberMarkus** [comment, marked as answer]
Glad I could help with your custom configuration file.

If you were satisfied with my free support, please consider [donating](https://github.com/sponsors/GruberMarkus) to this Open-Source project.
<br>
Dear businesses using Set-OutlookSignatures:
- Being Free and Open-Source Software, Set-OutlookSignatures can save you thousands or even tens of thousand Euros/US-Dollars per year in comparison to commercial software.  
Please consider [sponsoring this project](https://github.com/sponsors/GruberMarkus) to ensure continued support, testing and enhancements.
- Invest in the open-source projects you depend on. Contributors are working behind the scenes to make open-source better for everyone - give them the help and recognition they deserve.
- Sponsor the open-source software your team has built its business on. Fund the projects that make up your software supply chain to improve its performance, reliability, and stability.

**Reply by Imaginos** [reply]
I will ask my supply dept. for a donation !
It is worthy, for sure :)
